src/dash-table/components/Button/ExportButton/utils.tsx

```diff
@@ -0,0 +1,74 @@
+import * as R from 'ramda';
+import XLSX from 'xlsx';
+
+interface IMergeObject {
+    s: {r: number, c: number};
+    e: {r: number, c: number};
+}
+
+export function transformMultDimArray(array: (string | string[])[], maxLength: number): string[][] {
+    const newArray: string[][] = array.map(row => {
+        if (row instanceof Array && row.length < maxLength) {
+            return row.concat(Array(maxLength - row.length).fill(''));
+        }
+        if (maxLength === 0 || maxLength === 1) {
+            return [row];
+        }
+        if (row instanceof String || typeof(row) === 'string') {
+            return Array(maxLength).fill(row);
+        }
+        return row;
+    });
+    return newArray;
+}
+
+export function getMergeRanges(array: string[][]) {
+    let apiMergeArray: IMergeObject[] = [];
+    const iForEachOuter = R.addIndex<(string[]), void>(R.forEach);
+    const iForEachInner = R.addIndex<(string), void>(R.forEach);
+    iForEachOuter((row: string[], rIndex: number) => {
+        let dict: any = {};
+        iForEachInner((cell: string, cIndex: number) => {
+            if (!dict[cell]) {
+                dict[cell] = {s: {r: rIndex, c: cIndex}, e: {r: rIndex, c: cIndex }};
+            } else {
+                if (cIndex === (dict[cell].e.c + 1)) {
+                    dict[cell].e = {r: rIndex, c: cIndex};
+                } else {
+                    apiMergeArray.push(dict[cell]);
+                    dict[cell] = {s: {r: rIndex, c: cIndex}, e: {r: rIndex, c: cIndex }};
+                }
+            }
+        }, row);
+        const objectsToMerge: IMergeObject[] = Object.values(dict);
+        apiMergeArray = R.concat(apiMergeArray, objectsToMerge );
+    }, array);
+    return R.filter((item: IMergeObject) => item.s.c !== item.e.c || item.s.r !== item.e.r, apiMergeArray);
+}
+
+export function createWorkbook(ws: XLSX.WorkSheet, Heading: string[][], exportHeader: string) {
+    const wb = XLSX.utils.book_new();
+    XLSX.utils.book_append_sheet(wb, ws, 'SheetJS');
+    if (exportHeader === 'display') {
+        wb.Sheets.SheetJS['!merges'] = getMergeRanges(Heading);
+    }
+    return wb;
+}
+
+export function createWorksheet(Heading: string[][], data: any[], columnID: string[], exportHeader: string ) {
+    const ws = XLSX.utils.aoa_to_sheet(Heading);
+    if (exportHeader === 'display' || exportHeader === 'names' || exportHeader === 'none') {
+        XLSX.utils.sheet_add_json(ws, data, {
+            skipHeader: true,
+            origin: Heading.length
+        });
+    } else if (exportHeader === 'ids') {
+        XLSX.utils.sheet_add_json(ws, data, { header: columnID });
+    }
```

Comment:
Can you explain why columnID as headers vs. calculated headers need to be added differently to the worksheet?

Reply:
the object that contains headers can only accept a 1D array but calculated header is 2D array

Reply:
Ok, would there be a problem  if we were to also make the `ids` case a `string[][]` then we could simplify the worksheet creation / abstract it away from the table, and only need `headers: string[][] + data`?